rcl_action/include/rcl_action/action_client.h

```diff
@@ -0,0 +1,690 @@
+ * \return `RCL_RET_OK` if the action client was deinitialized successfully, or
+ * \return `RCL_RET_ACTION_CLIENT_INVALID` if the action client is invalid, or
+ * \return `RCL_RET_NODE_INVALID` if the node is invalid, or
+ * \return `RCL_RET_ERROR` if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_action_client_fini(rcl_action_client_t * action_client, rcl_node_t * node);
+
+/// Return the default action client options in a rcl_action_client_options_t.
+/**
+ * The defaults are:
+ *
+ * - qos = TODO(jacobperron): RFC where to define default? and what should it be?
+ * - allocator = rcl_get_default_allocator()
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_action_client_options_t
+rcl_action_client_get_default_options(void);
+
+/// Send a ROS goal using a rcl_action_client_t.
+/**
+ * This is a non-blocking call.
+ *
+ * The caller is responsible for ensuring that the type of `ros_goal_request`
+ * and the type associate with the client (via the type support) match.
+ * Passing a different type produces undefined behavior and cannot be checked
+ * by this function and therefore no deliberate error will occur.
+ *
+ * The ROS goal message given by the `ros_goal_request` void pointer is always
+ * owned by the calling code, but should remain constant during execution of this
+ * function. i.e. Before and after calling rcl_action_send_goal_request() the
+ * `ros_goal_request` message can change, but it cannot be changed during the call to
+ * rcl_action_send_goal_request().
+ * The same `ros_goal_request` can be passed to multiple calls of this function
+ * simultaneously, even if the action clients differ.
+ *
+ * This function is thread safe so long as access to both the rcl_action_client_t
+ * and the `ros_goal_request` are synchronized.
+ * That means that calling rcl_action_send_goal_request() from multiple threads is allowed,
+ * but calling rcl_action_send_goal_request() at the same time as non-thread safe action
+ * client functions is not, e.g. calling rcl_action_send_goal_request() and
+ * rcl_action_client_fini() concurrently is not allowed.
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | No
+ * Thread-Safe        | Yes [1]
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ * <i>[1] for unique pairs of clients and goals, see above for more</i>
+ *
+ * \param[in] action_client handle to the client that will make the goal request
+ * \param[in] ros_goal_request pointer to the ROS goal message
+ * \return `RCL_RET_OK` if the request was sent successfully, or
+ * \return `RCL_RET_INVALID_ARGUMENT` if any arguments are invalid, or
+ * \return `RCL_RET_ACTION_CLIENT_INVALID` if the client is invalid, or
+ * \return `RCL_RET_ERROR` if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_action_send_goal_request(
+  const rcl_action_client_t * action_client,
+  const void * ros_goal_request);
```

Comment:
@jacobperron as @sloretz suggested for `rcl_action_take_feedback()` below, may be useful to expose `rmw_request_id_t` if multiple goal requests are sent one after the other as in:

https://github.com/ros2/rcl/blob/5e3d4be72089a4c766b508877bc52a196bd7bead/rcl/include/rcl/service.h#L300-L303

Reply:
Perhaps, I was assuming that the goal ID would be present in the ROS message (`ros_goal_request`). The client could use the goal ID to distinguish between multiple responses.